### PR TITLE
Eldritch input parameters from JSON.

### LINKED ISSUE #119: [feature][eldritch] Eldritch receive variable parameters through the `eldritch_run` function
**Is your feature request related to a problem? Please describe.**
To allow users to pass parameters through Tavern tomes into imix the eldritch run function needs to accept parameters and pass them into the interpreter.

**Describe the solution you'd like**
- Add `tome_parameters` to the `eldritch_run` function in `eldritch/src/lib.rs`
- Receive `tome_parameters` as a JSON string.
- Use serde to deserialize the `tome_parameters` into a `Dict` or `HashMap`
- Pass the deserialized parameters into the GlobalsBuilder using the `get_eldritch` function.
- Reference parameters as `input_vars['cmd']`

**Describe alternatives you've considered**
N/a

**Additional context**
N/a

### PR DESCRIPTION
**Pull request by hulto**
#### What type of PR is this?
/kind feature
/kind eldritch-function

#### What this PR does / why we need it:
Add support for parameters in eldritch.
- Allows tomes to be reusable
- Tome inputs can be string, number, or list of strings.

#### Which issue(s) this PR fixes:
Fixes #119

### CODE REVIEW

File: implants/eldritch/src/lib.rs [hunk 1]
```diff
@@ -36,26 +38,65 @@ pub fn eldritch_run(tome_filename: String, tome_contents: String) -> anyhow::Res
...
+
     let globals = get_eldritch()?;
+
     let module: Module = Module::new();
 
+    let res: SmallMap<Value, Value> = SmallMap::new();
+    let mut input_vars: Dict = Dict::new(res);
+
+    let parsed: serde_json::Value = serde_json::from_str(&tome_params_str)?;
+    let obj: serde_json::Map<String, serde_json::Value> = parsed.as_object().unwrap().clone();
```
Comment (Cictrone) [inline implants/eldritch/src/lib.rs:54]: unwrap = nick sadge

Comment (Cictrone) [reply implants/eldritch/src/lib.rs:54]: nit: name var something like param_map

File: implants/eldritch/src/lib.rs [hunk 2]
```diff
@@ -36,26 +38,65 @@ pub fn eldritch_run(tome_filename: String, tome_contents: String) -> anyhow::Res
...
+    for (key, value) in obj.iter() {
+        let mut new_value: Value = Value::new_none();
+        let new_key = module.heap().alloc_str(&key);
+
+        if value.is_string() {
+            new_value = module.heap().alloc_str(value.as_str().unwrap()).to_value();
```
Comment (Cictrone) [inline implants/eldritch/src/lib.rs:60]: unwrap

File: implants/eldritch/src/lib.rs [hunk 3]
```diff
@@ -36,26 +38,65 @@ pub fn eldritch_run(tome_filename: String, tome_contents: String) -> anyhow::Res
...
+        }else if value.is_array() {
+            let mut tmp_list: Vec<Value> = Vec::new();
+            for sub_value in value.as_array().unwrap() {
```
Comment (Cictrone) [inline implants/eldritch/src/lib.rs:63]: unwrap

File: implants/eldritch/src/lib.rs [hunk 4]
```diff
@@ -36,26 +38,65 @@ pub fn eldritch_run(tome_filename: String, tome_contents: String) -> anyhow::Res
...
+                if sub_value.is_string() {
+                    tmp_list.push(module.heap().alloc_str(sub_value.as_str().unwrap()).to_value());
```
Comment (Cictrone) [inline implants/eldritch/src/lib.rs:65]: unwrap

File: implants/eldritch/src/lib.rs [hunk 5]
```diff
@@ -36,26 +38,65 @@ pub fn eldritch_run(tome_filename: String, tome_contents: String) -> anyhow::Res
...
+                Some(tmp_i64) => match tmp_i64.try_into() {
+                    Ok(tmp_i32) => tmp_i32,
+                    Err(_) => i32::MAX.into(),
+                },
+                None => i32::MAX.into(),
+            };
+            new_value = Value::new_int(tmp_value);
+        }
+
+        input_vars.insert_hashed(new_key.to_value().get_hashed().unwrap(), new_value);
```
Comment (Cictrone) [inline implants/eldritch/src/lib.rs:81]: unwrap

File: implants/eldritch/src/lib.rs [hunk 6]
```diff
@@ -36,26 +38,65 @@ pub fn eldritch_run(tome_filename: String, tome_contents: String) -> anyhow::Res
...
+    }
+
+    module.set("input_vars", input_vars.alloc_value(module.heap()));
```
Comment (Cictrone) [inline implants/eldritch/src/lib.rs:84]: in docs/ui its called params, why not here?

File: implants/eldritch/src/lib.rs [hunk 7]
```diff
@@ -36,26 +38,65 @@ pub fn eldritch_run(tome_filename: String, tome_contents: String) -> anyhow::Res
...
         Ok(eval_val) => eval_val,
         Err(eval_error) => return Err(anyhow::anyhow!("Eldritch eval_module failed:\n{}", eval_error)),
     };
 
-    // Ok(res.unpack_str().unwrap().to_string())
-    let res_str = match res.unpack_str() {
-        Some(res) => res.to_string(),
-        None => return Err(anyhow::anyhow!("Failed to unpack result as str")),
-    };
+    let res_str = res.to_str();
```
Comment (Cictrone) [inline implants/eldritch/src/lib.rs:92]: does this panic? if so maybe should return to monke

File: implants/imix/src/main.rs [hunk 8]
```diff
@@ -318,8 +320,9 @@ mod tests {
 
         let result = runtime.block_on(handle_exec_tome(test_tome_input)).unwrap();
 
+        println!("{:?}", result.clone());
         let mut bool_res = false;
-        assert_eq!(result.1, "".to_string());
+        // assert_eq!(result.1, "".to_string());
```
Comment (Cictrone) [inline implants/imix/src/main.rs:325]: delete comment

File: implants/eldritch/src/lib.rs [hunk 9]
```diff
@@ -87,4 +128,54 @@ dir(pivot) == ["arp_scan", "bind_proxy", "ncat", "port_forward", "port_scan", "s
...
+    #[test]
+    fn test_library_parameter_input_max_u64() -> anyhow::Result<()>{
+        // Create test script
+        let test_content = format!(r#"
+x = input_vars["number"] - 1
+x
+"#);
+        let param_string = format!("{{\"number\":{}}}", u64::MAX);
+        let test_res = eldritch_run("test.tome".to_string(), test_content, Some(param_string));
+        // println!("{:?}", test_res);
```
Comment (Cictrone) [inline implants/eldritch/src/lib.rs:164]: delete comment

**Review by Cictrone — CHANGES_REQUESTED** [review]
kyle and nick in the mooooooooooorning

File: implants/imix/src/main.rs [hunk 10]
```diff
@@ -318,8 +320,8 @@ mod tests {
 
         let result = runtime.block_on(handle_exec_tome(test_tome_input)).unwrap();
 
+        println!("{:?}", result.clone());
```
Comment (KCarretto) [inline implants/imix/src/main.rs:323]: We sure we always want to print this?

**Review by KCarretto — APPROVED** [review]
Stunning. Gorgeous. Beautiful. Breathtaking.

**Pull request merged**